builder/api/src/main/java/io/helidon/builder/api/Prototype.java

```diff
@@ -0,0 +1,414 @@
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.List;
+
+import io.helidon.common.HelidonServiceLoader;
+import io.helidon.common.config.Config;
+import io.helidon.common.config.ConfiguredProvider;
+import io.helidon.common.config.NamedService;
+
+/**
+ * Prototype is generated from prototype blueprint, as is considered the public API.
+ */
+public interface Prototype {
+    /**
+     * Terminating method of this builder that creates a prototype instance.
+     * If the object is a factory, it has a further method {@code build}, that produces the target runtime instance
+     * created from this builder or setup object instance.
+     *
+     * @param <BUILDER>   type of the builder. see {@link #me()}
+     * @param <PROTOTYPE> type of the prototype to be built
+     * @see Prototype.Factory#build()
+     */
+    interface Builder<BUILDER, PROTOTYPE> {
+        /**
+         * Create an instance of the {@link Prototype}. This method is available on
+         * all generated builders for {@link Prototype.Blueprint}.
+         *
+         * @return an instance of the setup object created from this builder.
+         */
+        PROTOTYPE buildPrototype();
+
+        /**
+         * Instance of this builder as the correct type.
+         *
+         * @return this instance typed to correct type
+         */
+        @SuppressWarnings("unchecked")
+        default BUILDER me() {
+            return (BUILDER) this;
+        }
+
+        /**
+         * Resolve an instance to a built instance (if it is a builder).
+         *
+         * @param instance instance to check
+         * @return the same instance if not a builder, or a built instance
+         * @param <T> type of the parameter
+         */
+        @SuppressWarnings("unchecked")
+        default <T> T resolveBuilder(T instance) {
+            if (instance instanceof Builder<?, ?> builder) {
+                return (T) builder.buildPrototype();
+            }
+            if (instance instanceof io.helidon.common.Builder<?, ?> builder) {
+                return (T) builder.build();
+            }
+            return instance;
+        }
+    }
+
+    /**
+     * Extension of {@link io.helidon.builder.api.Prototype.Builder} that supports configuration.
+     * If a blueprint is marked as {@code @Configured}, build will accept configuration.
+     *
+     * @param <BUILDER>   type of the builder
+     * @param <PROTOTYPE> type of the prototype to be built
+     */
+    interface ConfiguredBuilder<BUILDER, PROTOTYPE> extends Builder<BUILDER, PROTOTYPE> {
+        /**
+         * Update builder from configuration.
+         *
+         * @param config configuration to use
+         * @return updated builder instance
+         */
+        BUILDER config(Config config);
+
+        /**
+         * Discover services from configuration.
```

Comment:
Thinks loaded through `ServiceLoader` and Pico `Service` are two different concepts. 
If you want to support injection, you create singletons. If you want to load extensions through service loader, you use this approach.
In imperative world (and for low level extensibility of server, media supports etc.) `ServiceLoader` is the right approach, as we may not use pico at all in these places.

Reply:
It might be the right choice to use `ServiceLoader` for these services - that is more of an implementation detail.

However, it wouldn't be a stretch to consider that these service instances will (a) want to be used as injectable services by some other pico service, and (b) we would want to ensure the proper cardinality of instance creation as I mentioned also.